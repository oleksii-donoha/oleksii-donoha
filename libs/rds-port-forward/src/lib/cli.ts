import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';

import { Mediator } from './mediator.js';

/**
 * CLI argument that can be skipped when running the command repeatedly
 */
type SkippableArg = {
  skippable: true;
  value: string | undefined;
};

type RequiredArg = {
  skippable: false;
  value: string;
};

export type Arg = SkippableArg | RequiredArg;

// used to narrow down the argv type to the options that we care about
const argKeys = [
  'cluster',
  'service',
  'container',
  'db-host',
  'db-host-from-container-env',
  'port',
  'local-port',
] as const;

export type ArgKey = (typeof argKeys)[number];

/**
 * Handles CLI interface and operations on arguments
 */
export class CliManager {
  protected mediator: Mediator;

  constructor(argv: typeof process.argv, mediator: Mediator) {
    this.mediator = mediator;
    const parsedArgv = yargs()
      .scriptName('rds-port-forward')
      .options({
        cluster: {
          describe: 'Name of the ECS cluster where target resides',
          type: 'string',
        },
        service: {
          describe:
            'Name (fuzzy) of the service that hosts target task\nRecommended to use when dealing with big clusters with lots of tasks',
          type: 'string',
        },
        container: {
          describe:
            'Name (fuzzy) of the container that will be used to forward the port',
          type: 'string',
        },
        'db-host': {
          describe:
            'Hostname (or IP address) of the DB instance to which the local port will be forwarded',
          type: 'string',
        },
        'db-host-from-container-env': {
          describe: `Target container's environment variable whose value points to the DB hostname (or IP)`,
          type: 'string',
        },
        port: {
          describe: 'Remote port to forward traffic to',
          type: 'string',
        },
        'local-port': {
          describe: 'Port on your machine that will listen to requests',
          type: 'string',
        },
        verbose: {
          describe: 'Prints more logs for debugging',
          type: 'boolean',
        },
      })
      .conflicts('db-host', 'db-host-from-container-env')
      .parseSync(hideBin(argv));

    this.mediator.rawArgs = argKeys.reduce(
      (acc, argKey) => {
        acc[argKey] = parsedArgv[argKey];
        return acc;
      },
      {} as Pick<typeof parsedArgv, ArgKey>,
    );
    this.mediator.verbose = parsedArgv.verbose ?? false;
  }

  /**
   * Returns the processed args committed to the mediator object by resolvers
   */
  get equivalent() {
    if (
      !this.mediator.processedArgs ||
      Object.keys(this.mediator.processedArgs).length === 0
    ) {
      throw new Error('There is no collected argument data from resolvers');
    }
    return this.mediator.processedArgs;
  }

  /**
   * Formats CLI arguments that can be used for repeat invocations
   * @param format 'full' or 'only-required'
   * @returns formatted string
   */
  public formatCliArgs(format: 'full' | 'only-required') {
    const args = Object.entries(this.equivalent).reduce((acc, [key, arg]) => {
      if (!arg.value || (format === 'only-required' && arg.skippable))
        return acc;
      acc.push(`\t--${key} ${arg.value}`);
      return acc;
    }, [] as string[]);
    return args.join(' \\\n');
  }
}
