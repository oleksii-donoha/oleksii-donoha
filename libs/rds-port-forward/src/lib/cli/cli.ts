import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';

import { Mediator } from '../mediator.js';
import {
  ArgKey,
  argKeys,
  awsCliSpecificOptions,
  cliOptions,
  conflictingCliOptions,
  NO_RESOLVER_DATA,
  scriptName,
} from './uiText.js';

/**
 * CLI argument that can be skipped when running the command repeatedly
 */
type SkippableArg = {
  skippable: true;
  value: string | undefined;
};

type RequiredArg = {
  skippable: false;
  value: string;
};

export type Arg = SkippableArg | RequiredArg;

// used to narrow down the argv type to the options that we care about

/**
 * Handles CLI interface and operations on arguments
 */
export class CliManager {
  protected mediator: Mediator;

  constructor(argv: typeof process.argv, mediator: Mediator) {
    this.mediator = mediator;
    const yargsInstance = yargs();
    const parsedArgv = yargsInstance
      .scriptName(scriptName)
      .options(cliOptions)
      .conflicts(...conflictingCliOptions)
      .version(false)
      .wrap(yargsInstance.terminalWidth())
      .parseSync(hideBin(argv));

    this.mediator.rawArgs = argKeys.reduce(
      (acc, argKey) => {
        acc[argKey] = parsedArgv[argKey];
        return acc;
      },
      {} as Pick<typeof parsedArgv, ArgKey>,
    );
    this.mediator.verbose = parsedArgv.verbose ?? false;
    for (const arg of awsCliSpecificOptions) {
      if (parsedArgv[arg]) {
        this.mediator.processedArgs[arg] = {
          skippable: false,
          value: parsedArgv[arg],
        };
      }
    }
  }

  /**
   * Returns the processed args committed to the mediator object by resolvers
   */
  get equivalent() {
    if (
      !this.mediator.processedArgs ||
      Object.keys(this.mediator.processedArgs).length === 0
    ) {
      throw new Error(NO_RESOLVER_DATA);
    }
    return this.mediator.processedArgs;
  }

  /**
   * Formats CLI arguments that can be used for repeat invocations
   * @param format 'full' or 'only-required'
   * @returns formatted string
   */
  public formatCliArgs(format: 'full' | 'only-required') {
    const args = Object.entries(this.equivalent).reduce((acc, [key, arg]) => {
      if (!arg.value || (format === 'only-required' && arg.skippable))
        return acc;
      acc.push(`\t--${key} ${arg.value}`);
      return acc;
    }, [] as string[]);
    return args.join(' \\\n');
  }
}
