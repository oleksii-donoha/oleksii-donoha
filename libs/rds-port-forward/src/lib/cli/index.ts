export * from './cli.js';
export { type ArgKey } from './uiText.js';
