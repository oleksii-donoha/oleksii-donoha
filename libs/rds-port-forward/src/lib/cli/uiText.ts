export const argKeys = [
  'cluster',
  'service',
  'container',
  'db-host',
  'db-host-from-container-env',
  'port',
  'local-port',
  'profile',
  'region',
] as const;

export type ArgKey = (typeof argKeys)[number];

export const cliOptions: {
  [K in ArgKey | 'verbose']: {
    describe: string;
    type: K extends 'verbose' ? 'boolean' : 'string';
  };
} = {
  cluster: {
    describe: 'Name of the ECS cluster where target resides',
    type: 'string',
  },
  service: {
    describe:
      'Name (fuzzy) of the service that hosts target task\nRecommended to use when dealing with big clusters with lots of tasks',
    type: 'string',
  },
  container: {
    describe:
      'Name (fuzzy) of the container that will be used to forward the port',
    type: 'string',
  },
  'db-host': {
    describe:
      'Hostname (or IP address) of the DB instance to which the local port will be forwarded',
    type: 'string',
  },
  'db-host-from-container-env': {
    describe:
      "Target container's environment variable whose value points to the DB hostname (or IP)",
    type: 'string',
  },
  port: {
    describe: 'Remote port to forward traffic to',
    type: 'string',
  },
  'local-port': {
    describe: 'Port on your machine that will listen to requests',
    type: 'string',
  },
  verbose: {
    describe: 'Prints more logs for debugging',
    type: 'boolean',
  },
  profile: {
    describe: 'AWS CLI profile to use',
    type: 'string',
  },
  region: {
    describe: 'AWS region for the request',
    type: 'string',
  },
};

export const conflictingCliOptions = [
  'db-host',
  'db-host-from-container-env',
] as const;

export const awsCliSpecificOptions = ['profile', 'region'] as const;

export const scriptName = 'rds-port-forward';

export const NO_RESOLVER_DATA =
  'There is no collected argument data from resolvers';
