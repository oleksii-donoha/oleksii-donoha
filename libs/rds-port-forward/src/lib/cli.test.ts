import { CliManager, CliOptionType } from './cli.js';
import { Logger } from 'winston';
import { vi, describe, it, expect, beforeEach } from 'vitest';

describe('CliManager', () => {
  let logger: Logger;
  let argv: string[];

  beforeEach(() => {
    logger = {
      debug: vi.fn(),
      error: vi.fn(),
    } as unknown as Logger;
  });

  describe('constructor', () => {
    it('should parse CLI arguments and initialize incoming args', () => {
      argv = ['node', 'script.js', '--cluster', 'test-cluster'];
      const cliManager = new CliManager(logger, argv);

      expect(cliManager['incoming']).toEqual({
        cluster: 'test-cluster',
        service: undefined,
        container: undefined,
        'db-host': undefined,
        'db-host-from-container-env': undefined,
      });
    });
  });

  describe('equivalent', () => {
    it('should throw an error if equivalent is not set', () => {
      argv = ['node', 'script.js', '--cluster', 'test-cluster'];
      const cliManager = new CliManager(logger, argv);

      expect(() => cliManager.equivalent).toThrow(
        'The equivalent CLI args are not set. Did you forget to set them while processing the target?'
      );
    });

    it('should return the equivalent CLI args if set', () => {
      argv = ['node', 'script.js', '--cluster', 'test-cluster'];
      const cliManager = new CliManager(logger, argv);
      cliManager.markCliOptionAs(
        CliOptionType.Skippable,
        'cluster',
        'test-cluster'
      );

      expect(cliManager.equivalent).toEqual({
        cluster: { value: 'test-cluster', skippable: true },
      });
    });

    it('equivalent args should differ from initial ones when overwritten', () => {
      argv = ['node', 'script.js', '--service', 'test-service'];
      const cliManager = new CliManager(logger, argv);
      cliManager.markCliOptionAs(
        CliOptionType.Required,
        'service',
        'other-service'
      );

      expect(cliManager.equivalent).toEqual({
        service: { value: 'other-service', skippable: false },
      });
    });
  });
  describe('markAs', () => {
    it('should correctly set the skippable flag based on the option type', () => {
      argv = ['node', 'script.js', '--cluster', 'test-cluster'];
      const cliManager = new CliManager(logger, argv);
      cliManager.markCliOptionAs(
        CliOptionType.Skippable,
        'cluster',
        'test-cluster'
      );
      cliManager.markCliOptionAs(
        CliOptionType.Required,
        'container',
        'test-container'
      );

      expect(cliManager.equivalent).toEqual({
        container: { value: 'test-container', skippable: false },
        cluster: { value: 'test-cluster', skippable: true },
      });
    });
  });
});
