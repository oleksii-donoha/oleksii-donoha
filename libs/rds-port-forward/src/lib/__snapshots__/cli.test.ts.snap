// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CliManager > formatCliArgs > should format all CLI arguments when format is "full" 1`] = `
"	--cluster test-cluster \\
	--service test-service \\
	--port 5432"
`;

exports[`CliManager > formatCliArgs > should format only required CLI arguments when format is "only-required" 1`] = `
"	--cluster test-cluster \\
	--port 5432"
`;

exports[`CliManager > formatCliArgs > should skip arguments with undefined values 1`] = `"	--service test-service"`;
